### DISCUSSION TITLE
How to copy a cell / paste into a cell?

### DISCUSSION THREAD

**Discussion by jvns** [saulpw/visidata#983] [Q&A]
I can see from the docs how to copy/paste a row but I can't figure out to copy just the contents of a cell or how to paste the contents of my clipboard into a cell.

**Comment by ajkerrigan** [comment]
Hiya! You can use `zy`/`zp` to copy/paste a cell within VisiData, or `zY` to copy a cell to the system clipboard. The default binding for pasting from the system clipboard is mouse button 2.

In case it's useful to you or anyone else (I forget certain bindings _way too often_, perhaps I'm not alone? :grin: ), this will show all key bindings related to copy/paste:

1. Open keyboard shortcut help with `z` `^H` (or `z` `F1`)
2. Move to the `longname` column
3. Hit `|` and search for `copy|paste` to select rows where the command name contains copy or paste
4. Use `"` to open a new sheet with only the selected rows

By default you should see something like this:

```
 sheet              | longname           ║ keystrokes       | description
 BaseSheet          | syscopy-col        ║                 ⌀| yank (copy) current column values to system clipboard (using opt…
 TableSheet         | copy-cell          ║ zy               | yank (copy) current cell to clipboard
 TableSheet         | copy-cells         ║ gzy              | yank (copy) contents of current column for selected rows to clip…
 TableSheet         | copy-row           ║ y                | yank (copy) current row to clipboard
 TableSheet         | copy-selected      ║ gy               | yank (copy) selected rows to clipboard
 TableSheet         | paste-after        ║ p                | paste clipboard rows after current row
 TableSheet         | paste-before       ║ P                | paste clipboard rows before current row
 TableSheet         | paste-cell         ║ zp               | set contents of current cell to last clipboard value
 TableSheet         | syscopy-cell       ║ zY               | yank (copy) current cell to system clipboard (using options.clip…
 TableSheet         | syscopy-cells      ║ gzY              | yank (copy) contents of current column from selected rows to sys…
 TableSheet         | syscopy-row        ║ Y                | yank (copy) current row to system clipboard (using options.clipb…
 TableSheet         | syscopy-selected   ║ gY               | yank (copy) selected rows to system clipboard (using options.cli…
 TableSheet         | syspaste-cells     ║ BUTTON2_RELEASED | paste into VisiData from system clipboard
```

Which is a pretty handy reference! Especially if you've defined some of your own bindings and subsequently forgotten them. :innocent: . Copy/paste generally follows the typical VisiData pattern of "g goes big, z goes small", but there are a bunch of subtle variations to keep track of and it's nice to have help :).

**Comment by saulpw** [comment]
Maybe we should have an actual keybinding for `syspaste-cells`, so a mouse-averse user doesn't have to use Space to execute the command by longname. What do you think about `z Shift+P`?  It's a little asymmetric (and makes it seem like `Shift+P` would paste a row from the system clipboard, but actually it pastes the internal clipboard row before the cursor, a la vim), but maybe that's okay?  I don't know what other binding would make sense, which is why it doesn't have one yet.

**Reply by ajkerrigan** [reply]
As a somewhat-mouse-averse user myself, I use `zP` as `syspaste-cells`! :) I like the symmetry of `zY`/`zP` even if it seems a little out of step with the other row-focused paste bindings :shrug:

**Reply by ghost** [reply]
The way that vim manages this is by having named clipboard registers. The default register is the internal register (like visidata), but you can optionally interact with the system clipboard by prefixing `"*` before the clipboard command.

**Reply by frosencrantz** [reply]
Saul has granted this wish in https://github.com/saulpw/visidata/issues/990:
`Okay, I've added a binding for z Shift+P, and a new command gz Shift+P to paste onto selected rows.`
